File: daily/2025/1504_Count_Submatrices_With_All_Ones.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int m = mat.size();
        int n = mat[0].size();
        vector<vector<int>> height(m, vector<int>(n, 0));
        for (int j = 0; j < n; j++) height[0][j] = mat[0][j];
        for (int i = 1; i < m; i++){
            for (int j = 0; j < n; j++){
                height[i][j] = (mat[i][j] == 0)? 0 : height[i-1][j] + 1;
            }
        }
        int ans = 0;
        for (int i = 0; i < m; i++){
            stack<vector<int>> s;
            s.push({-1, 0, -1});
            for (int j = 0; j < n; j++){
                int curr_height = height[i][j];
                while(s.top()[2] >= curr_height) s.pop();
                int k = s.top()[0];
                int acc = s.top()[1];
                int curr_acc = acc + (j - k) * curr_height;
                s.push({j, curr_acc, curr_height});
                ans += curr_acc;
            }
        }
        return ans;
    }
};

int main(){
    vector<vector<int>> mat = {{1,0,1},{1,1,0},{1,1,0}};
    // vector<vector<int>> mat = {{0,1,1,0},{0,1,1,1},{1,1,1,0}};
    // vector<vector<int>> mat = {{1,1,1,1,0},{1,0,0,1,0},{0,0,1,0,1},{0,1,0,0,0}};
    Solution S;
    cout << S.numSubmat(mat) << endl;
    return 0;
}

File: daily_python/1504_Count_Submatrices_With_All_Ones.py
import os
import sys
from typing import List

class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        m = len(mat)
        n = len(mat[0])
        height = [0] * n
        ans = 0
        for row in range(m):
            height = [0 if b == 0 else h + 1 for (b, h) in zip(mat[row], height)]
            # (col, acc, height)
            stack = [(-1, 0, -1)]
            for col in range(n):
                curr_height = height[col]
                while curr_height <= stack[-1][2]:
                    stack.pop()
                prev_col = stack[-1][0]
                prev_acc = stack[-1][1]
                curr_acc = (col - prev_col) * curr_height + prev_acc
                stack.append((col, curr_acc, curr_height))
                ans += curr_acc
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            mat = [list(map(int, row.split(','))) for row in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append(mat)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for mat in dataset:
        results.append(solution.numSubmat(mat))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')